vm/src/builtins.rs

```diff
@@ -730,100 +730,80 @@ fn builtin_zip(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
-    ctx.set_attr(&py_mod, "classmethod", ctx.classmethod_type());
-    ctx.set_attr(&py_mod, "compile", ctx.new_rustfunc(builtin_compile));
-    ctx.set_attr(&py_mod, "complex", ctx.complex_type());
-    ctx.set_attr(&py_mod, "delattr", ctx.new_rustfunc(builtin_delattr));
-    ctx.set_attr(&py_mod, "dict", ctx.dict_type());
-    ctx.set_attr(&py_mod, "divmod", ctx.new_rustfunc(builtin_divmod));
-    ctx.set_attr(&py_mod, "dir", ctx.new_rustfunc(builtin_dir));
-    ctx.set_attr(&py_mod, "enumerate", ctx.new_rustfunc(builtin_enumerate));
-    ctx.set_attr(&py_mod, "eval", ctx.new_rustfunc(builtin_eval));
-    ctx.set_attr(&py_mod, "exec", ctx.new_rustfunc(builtin_exec));
-    ctx.set_attr(&py_mod, "float", ctx.float_type());
-    ctx.set_attr(&py_mod, "frozenset", ctx.frozenset_type());
-    ctx.set_attr(&py_mod, "filter", ctx.new_rustfunc(builtin_filter));
-    ctx.set_attr(&py_mod, "format", ctx.new_rustfunc(builtin_format));
-    ctx.set_attr(&py_mod, "getattr", ctx.new_rustfunc(builtin_getattr));
-    ctx.set_attr(&py_mod, "hasattr", ctx.new_rustfunc(builtin_hasattr));
-    ctx.set_attr(&py_mod, "hash", ctx.new_rustfunc(builtin_hash));
-    ctx.set_attr(&py_mod, "hex", ctx.new_rustfunc(builtin_hex));
-    ctx.set_attr(&py_mod, "id", ctx.new_rustfunc(builtin_id));
-    ctx.set_attr(&py_mod, "int", ctx.int_type());
-    ctx.set_attr(&py_mod, "isinstance", ctx.new_rustfunc(builtin_isinstance));
-    ctx.set_attr(&py_mod, "issubclass", ctx.new_rustfunc(builtin_issubclass));
-    ctx.set_attr(&py_mod, "iter", ctx.new_rustfunc(builtin_iter));
-    ctx.set_attr(&py_mod, "len", ctx.new_rustfunc(builtin_len));
-    ctx.set_attr(&py_mod, "list", ctx.list_type());
-    ctx.set_attr(&py_mod, "locals", ctx.new_rustfunc(builtin_locals));
-    ctx.set_attr(&py_mod, "map", ctx.new_rustfunc(builtin_map));
-    ctx.set_attr(&py_mod, "max", ctx.new_rustfunc(builtin_max));
-    ctx.set_attr(&py_mod, "min", ctx.new_rustfunc(builtin_min));
-    ctx.set_attr(&py_mod, "object", ctx.object());
-    ctx.set_attr(&py_mod, "oct", ctx.new_rustfunc(builtin_oct));
-    ctx.set_attr(&py_mod, "ord", ctx.new_rustfunc(builtin_ord));
-    ctx.set_attr(&py_mod, "next", ctx.new_rustfunc(builtin_next));
-    ctx.set_attr(&py_mod, "pow", ctx.new_rustfunc(builtin_pow));
-    ctx.set_attr(&py_mod, "print", ctx.new_rustfunc(builtin_print));
-    ctx.set_attr(&py_mod, "property", ctx.property_type());
-    ctx.set_attr(&py_mod, "range", ctx.new_rustfunc(builtin_range));
-    ctx.set_attr(&py_mod, "repr", ctx.new_rustfunc(builtin_repr));
-    ctx.set_attr(&py_mod, "set", ctx.set_type());
-    ctx.set_attr(&py_mod, "setattr", ctx.new_rustfunc(builtin_setattr));
-    ctx.set_attr(&py_mod, "staticmethod", ctx.staticmethod_type());
-    ctx.set_attr(&py_mod, "str", ctx.str_type());
-    ctx.set_attr(&py_mod, "sum", ctx.new_rustfunc(builtin_sum));
+pub fn mk_module(ctx: &PyContext) -> PyObjectRef {
+    let py_mod = py_item!(ctx, mod __builtins__ {
```

Comment:
Thanks for your contribution! At first sight, I am unsure whether this makes the code more easy to understand / maintain. To me it first looks more obscure then before, but that is off course because I'm used to the previous way of doing things. I'm curious to other opinions!

Reply:
I do see what you mean. My motivation for doing this was to have a more declarative syntax for creating modules and to be able to separate the different portions in a more concrete way than `let x = ctx.new_x()` and `ctx.set_attr(py_mod, "x", x)`. Do you have any qualms with the syntax itself, or just the macro?

Reply:
My qualms with this change are the following:
- `py_item` is too vague, it is better to be more concrete, so that is clear what is meant.
- I prefer functions over macros in general. Did you for example check code size before and after your change? Also in rust, macros can be very ambiguous. A function is more clear in my opinion. I would say, use macros when you have to, functions otherwise.
- The fact that we need to set macro limit to 128 might be an indication that it is probably too much macro.

That being said, I would like feedback from others about this, if more people are in favor, I'm all for it!

Reply:
Btw, thank you very much for your contributions, I understand your point with first assigning `x` and then using `set_attr` again.

Reply:
Upon reflection, I realize that you probably meant code size before macro expansion, i.e. LOC, but FWIW, [here's](https://gist.github.com/coolreader18/224cfa96ad8b7e2cb62dd12bb31efdfa) a comparison of `io.rs` without macros and what it looks like with macros after expansion.

Reply:
If this doesn't get merged, it does seem like a more organized pattern to do something like:
```rust
let string_io_type = {
   // assemble string_io
};
ctx.set_attr(&py_mod, "StringIO", string_io_type);
```

Reply:
Lines of source code compare with no macros having 11 LOC without empty lines and 15 with, and macros having ten with no empty lines necessary to organize the code.

Reply:
What I meant by code size was the resulting binary size. Since macros expand code, the use of macros might result in bigger binaries.